### ISSUE TITLE
Versioning for repos that are not tagged with named release branches

### ISSUE THREAD

**Issue by kdmccormick** [openedx/docs.openedx.org#391]
Read https://github.com/openedx/axim-engineering/issues/924#issuecomment-1765127544

Many repos don't get tagged in named releases because they are pip-installed into repos which are tagged.

How should that work with the version-picker in those repos' docs projects?

**Comment by feanil** [comment]
My gut feeling is that for libraries, we should tag docs for semantic version but we should also tag the relevant version with the openedx named release tag.  The reason for this is that from a consumer perspective, if you're writing code against or trying to user features from a named release, you shouldn't have to know what numeric version maps to the named release you care about.

For example: If there is a feature in the ORA repo that is available in Quince, and you want to look at the relevant docs, you should be able to pull up the quince docs for ORA whether that's version 6.5 or 7.1 in the platform isn't really useful.

This is definitely not how we've been doing things but I believe it's the direction we should be moving in.  In terms of automation, I think it's possible to update the release tagging script to take into account more catalog-info and for libraries, it could tag the latest released version with the relevant tree name as well.  There's still some edge cases to work out there but I think the solution is tractable.

**Comment by kdmccormick** [comment]
I agree in spirit, but I'm still not sure how to practically achieve this.

Imagine we have package "bork-lib" with these versions:

* main (latest code)
* 6.3 (latest release)
* 6.2  <-- edx-platform pins this one
* 6.1 <-- course-discovery pins this one

When it comes to tag the release, which version should we tag? I think the answer depends on what the consumer is doing. If they are hacking on discovery, they'll want to see 6.1; for edx-platform, 6.2; if they are developing on a new repo, they'll want 6.3. Our standard tagging process would tag main.

**Comment by kdmccormick** [comment]
> For example: If there is a feature in the ORA repo that is available in Quince, and you want to look at the relevant docs, you should be able to pull up the quince docs for ORA whether that's version 6.5 or 7.1 in the platform isn't really useful.

I read this wrong the first time. Based on what you're saying, we'd just want to choose the latest release that exists at cut time, `bork-lib==6.3` in the example above.

It'd be a tooling change, but not a very difficult one. I could get on board with that.

**Comment by nedbat** [comment]
We stopped branching and tagging dependent repos in 2016 (https://github.com/openedx/repo-tools/commit/ffead27), so Gingko was the last release that had them.

**Comment by mariajgrimaldi** [comment]
I found myself needing release-specific docs for events and filters, so developers know which events/filters are available for each release. Can those repos be included in the release process? I'm currently generating the docs for the open-release/*.master manually, but I'd like to know the state of this ticket to understand how maintainable this is in the long run.

**Comment by kdmccormick** [comment]
Coming to this ticket after all this time, I think the sensible solution would be something like this:

* For any given build of docs.openedx.org, we look at edx-platform's requirement pins, and use those to determine which package versions to pull in.
* We would do this both for docs.openedx.org master, as well as for previous named releases.
* Example:
  * To build master of docs.openedx.org, we would look at edx-platform's package metadata on master to determine that we should link to the docs build of openedx-filters v1.12.0.
  * To build redwood of docs.openedx.org, we would look at edx-platform's package metadata on open-release/redwood.master to determine that we should link to the docs build of openedx-filters v1.8.1.
* Open question: How would we ensure that the correct version of openedx-filters (and other packages) are build and available to be linked to?

@feanil , do you think that this ^ is both reasonable approach, and something that a core contributor would feasibly be able to orchestrate in docs.openedx.org?

**Comment by mariajgrimaldi** [comment]
I manually built versions for open-release/sumac.master for both libraries, using the latest versions available at the time. Would it be possible to consider using these release branches instead of programmatically determining the versions for the release? Or do we want to avoid tagging these pip installed libs?

**Comment by mariajgrimaldi** [comment]
Thank you for the detailed explanation, @kdmccormick. Yes, please go ahead. It'd be useful to establish a process for these kinds of requests.

**Comment by kdmccormick** [comment]
After talking about this with @feanil , we believe that tagging packages with named releases would introduce a lot of process complexity. Instead, automatically linking named-release-version docs on docs.openedx.org to the correct numbered-version package docs is the way to go. We'll need to build some tooling into edx-platform and docs.openedx.org in order to look up the pip-installed versions and generates the right links.
